### ISSUE TITLE
Use TopologyAwareHints in ebpf kube-proxy

### ISSUE THREAD

**Issue by HemantNegi** [projectcalico/calico#7041]
We are running calico wiht eBPF dataplane enabled and external kube-proxy disabled in our kuberntes clusters on aws. In this mode calico imports upstream kube-proxy logic and runs it with-in itself. We want to optimize our inter-AZ network costs and came across this solution offered by k8s [Topology aware hints](https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/). This feature is still in Beta and we have to turn it on via a feature-gate in kube-proxy. Now the problem is calico does not allow us to configure feature gates for kube-proxy. 

The required feature gate to be enabled is `TopologyAwareHints=true`

## Expected Behavior
Best will be if we can configure kube-proxy feature gates via some flag or environment variable.

## Current Behavior
This feature is disabled in kube-proxy by default.

## Possible Solution
- We can provide a featureGates flag in felix configuration that lests us enable disable features in kube-proxy.
- We can just enable this feature by adding 
`_ = utilfeature.DefaultMutableFeatureGate.Set("TopologyAwareHints=true")` here - https://github.com/projectcalico/calico/blob/master/felix/bpf/proxy/kube-proxy.go#L34

## Context
This feature in kuebernets allows the affinity for service endpoints so that they prefer other pods running in the same availability zone. This saves a lot of network cost and overall latency as traffic remains in same region 

* Calico version: v3.24.5
* Orchestrator version (e.g. kubernetes, mesos, rkt): Kubernetes 1.21
* Operating System and version: Linux

**Comment by seh** [comment]
There was some previous coverage of this topic [in the "kubernetes" channel](https://calicousers.slack.com/archives/C0BCA117T/p1653590202923409) of the "Calico Users" Slack workspaces.

**Comment by lwr20** [comment]
pinging @tomastigera

**Comment by tomastigera** [comment]
> Calico eBPF dataplane with built-in kube-proxy, must allow passing feature gates to kube-proxy.

It is not that simple. We also need to act on the extra information ;-)

**Comment by StevenTigera** [comment]
Hi @HemantNegi - could I please ask if you have been able to configure [Topology Aware Hints](https://kubernetes.io/docs/concepts/services-networking/_print/#pg-374e5c954990aec58a0797adc70a5039) [via feature gate] with Calico using Installation `linuxDataplane: Iptables` and if so what method(s) did you employ to get set up accordingly?  Thanks a lot

**Comment by HemantNegi** [comment]
> I please ask if you have been able to configure [](https://kubernetes.io/docs/concepts/services-networking/_print/#pg-374e5c954990aec58a0797adc70a5039)

@StevenTigera Unfortunately No. We ended up disabling eBPF and enabling kube-porxy back for the timebeing. This setup is working fine as expected. and we are planning to roll it up on production clsuters.

**Comment by tomastigera** [comment]
> > I please ask if you have been able to configure
> 
> @StevenTigera Unfortunately No. We ended up disabling eBPF and enabling kube-porxy back for the timebeing. This setup is working fine as expected. and we are planning to roll it up on production clusters.

@HemantNegi to reiterate on the question above, it was specifically about setting it up in iptables (not ebpf) mode in your environment. @StevenTigera is working on adding that feature to the ebpf dataplane.

**Comment by StevenTigera** [comment]
Hi @HemantNegi - just to clarify above RE: "This setup is working fine as expected" - does this mean you were able to use TopologyAwareHints by annotating your service(s) like this `service.kubernetes.io/topology-aware-hints: auto` thus the corresponding `EndpointSlice` resources auto-populated the endpoints `hints`?

If so then could you pls advise if needed to set feature gates in order to achieve this.  I read in the K8s docs that enabling features via feature gates may involve adding `--feature-gates=TopologyAwareHints=true` to  `/etc/kubernetes/manifests/kube-controller-manager.yaml` file on the master node and to `/etc/systemd/system/kubelet.service.d` file on each worker node, however this made no difference - thanks

**Comment by HemantNegi** [comment]
@StevenTigera Yes, we added the annotation (`service.kubernetes.io/topology-aware-hints: auto`) on our services and the corresponding `EndpointSlice`'s were populated with `hints` section.

To do this we enabled that feature gate on `api-server` and `kube-proxy`. Yeah the k8s doc says that kubelet and controller-manager can accept those values. But it's not required. 

The primary actors of this feature are `api server` and  `kube-proxy`.
